### ISSUE TITLE
vim.ui.select "blocking" version that yields result similar to inputlist()

### ISSUE THREAD

**Issue by lervag** [neovim/neovim#24632]
### Problem

`vim.ui.select` is based on the inversion of control pattern and takes as an argument the callback that will handle the selected element. This is clearly a very useful function!

But I was thinking it would also be useful with the blocking variant, perhaps called `vim.ui.select_blocking`, that will instead return the selected element and return control to the calling function. I know we have `inputlist()`, with which I can do this:

```lua
local foo = function()
  local list = {
    "one",
    "two",
    "three",
  }

  local idx = vim.fn.inputlist(list)

  print("You chose: " .. list[idx])
end
```

However, I would argue that `inputlist()` is quite restricted with regards to allowing modern UIs. With `vim.ui.select`, we are able to give the user quite a lot of flexibility. Both with regards to things like adding a formatter function and on allowing different implementations/backends (e.g. fzf, Telescope). Further, `inputlist()` really "feels" old fashioned. It does not work well with things like noice.nvim. I could go on, but I believe most of you understand what I'm talking about here.

### Expected behavior

I would like to do things like this:

```lua
local foo = function()
  local list = {
    "one",
    "two",
    "three",
  }

  local selection = vim.ui.select_blocking(list, {
      prompt =  "please select:",
      format_item =  function(i) return "→ " .. i end,
  })

  print("You chose: " .. selection)
end
```

Where `select_blocking` can have a useful default implementation that can be overridden by users that want to use e.g. things like fzf, Telescope or similar as backends.

**Comment by lervag** [comment]
By the way: If the community should agree that this would be useful, then I would also think the same "logic" applies for building the corresponding blocking variants of `vim.ui.input` and `vim.ui.confirm`. If so, I can open similar feature requests for these. But I'll hold back until we've had a discussion on the `select` behaviour first.

**Comment by clason** [comment]
I think that could be done with `sync = true` keyword, so no need for separate functions. (Of course, whether the overriding UIs respect that is another question entirely...)

**Comment by lewis6991** [comment]
Or luv style where omitting a callback infers sync.

**Comment by mfussenegger** [comment]
> Where select_blocking can have a useful default implementation that can be overridden by users that want to use e.g. things like fzf, Telescope or similar as backends.

It's afaik impossible to implement a version using fzf or telescope that's *blocking* the main loop. The main loop has to run in order to display the window and process input. 

The way to go is to use coroutine's if you want to avoid callback nesting. See also https://github.com/neovim/neovim/issues/19624

**Comment by lervag** [comment]
> > Where select_blocking can have a useful default implementation that can be overridden by users that want to use e.g. things like fzf, Telescope or similar as backends.
> 
> It's afaik impossible to implement a version using fzf or telescope that's _blocking_ the main loop. The main loop has to run in order to display the window and process input.

No, it is _not_ impossible. And IMHO clearly so: More or less all of the "classic" selector plugins were blocking (e.g. CtrlP).

We have all the building blocks to build a blocking selector. I built one for wiki.vim - the source is [here](https://github.com/lervag/wiki.vim/blob/04459625f1b6faa2b02ce763fc6fbc24d4a51f21/autoload/wiki/ui/nvim.vim#L90-L143). I'm not saying it's very good, my point is only to say that this is indeed possible.

For Telescope, it _would_ require some work that perhaps no one is interested in doing. But for fzf, I think it is fair to point out that the first fzf integrations in Vim was always blocking (but clearly very much of a hack!).

> The way to go is to use coroutine's if you want to avoid callback nesting. See also #19624

If we use coroutine's then we don't need the blocking variant. However, I believe this means we need to be able to control the coroutine's from the caller. This means the concept becomes impossible to work with in Vimscript, but it also means it will be much harder to use also in Lua scripts.

> > I think that could be done with sync = true keyword (or omitted on_choice function?), so no need for separate functions. (Of course, whether the overriding UIs respect that is another question entirely...)
> 
> > Or luv style where omitting a callback infers sync.
> 
> Even if it would be possible to implement a blocking style, adding new options or changing signatures is unfortunately also problematic - see discussions in #21681 and #20053

I believe you might be right about this one, though - I don't quite see how to properly combine the blocking style with the non-blocking in a single function. I think the only way would be to have an outer interface that wraps both variants. And this would be a major breaking change.

Personally, I'm fine with having a new name for the blocking style.

**Comment by mfussenegger** [comment]
Oh right, I didn't think of blocking via `getchar`.


In that case, a way forward could be to:

- Switch `vim.ui.select` to a pull/provider model like health checks. It would discover implementations, and use them if they have the required capabilities. Otherwise it has a fallback implementation.
   E.g. there could be a `vim/provider/ui` namespace for plugins, where a plugin can put a module into `vim/provider/ui.lua` that returns supported functions and their capabilities.

This would allow extending the function with new options as needed.

- Support luv style as Lewis suggested

- In another release, add metatable to `vim.ui` that warns overriding functions is deprecated

- In another release after ^, turn `vim.ui` into a frozen table.

**Comment by justinmk** [comment]
Yes https://github.com/neovim/neovim/issues/19624 seems very relevant, there should be a general way to "await" async functionality.

> E.g. there could be a vim/provider/ui namespace for plugins

"pull/provider model" is generally the right way, but can't we avoid the explicit `provider` namespace? With something like `on_func` #22598 I am convinced that _any_ function can be transformed into a "provider" model, we don't need to explicitly organize it for special functions. 

Providers can "publish" their capabilities by accepting a special arg, e.g. calling `vim.ui.select({capabilities=true})` gets the capabilities. Or this could be registered when `on_func` is called.

**Comment by lervag** [comment]
> In that case, a way forward could be to: …

Yes, your suggestions make a lot of sense to me! And I agree with your argument that adding a new function is suboptimal.

**Comment by mfussenegger** [comment]
> Providers can "publish" their capabilities by accepting a special arg, e.g. calling vim.ui.select({capabilities=true}) gets the capabilities. Or this could be registered when on_func is called.

I think we need at least one level of indirection between consumer/implementation to be able to add fallback implementations transparently.
Otherwise we'd push the responsibility to check the capabilities to all consumers.


Instead of:

PluginA, B, C, D:

```
local capabilities = vim.ui.select({ capabilities = true })
if capabilities.xy then
    ...
else
    -- custom falback implementation
end
```


We'd have:

PluginA, B, C, D:

```
vim.ui.select(...)
```

And in neovim the implementation would lookup providers, check it's capabilities, and use a fallback implementation if the requested capabilities are missing.
We'd need to have a spot for this provider lookup and fallback logic. If users continue to override `vim.ui.select` directly, or via `on_func`, we wouldn't have that.

**Comment by justinmk** [comment]
> We'd need to have a spot for this provider lookup and fallback logic. If users continue to override `vim.ui.select` directly, or via `on_func`, we wouldn't have that.

Couldn't `on_func` (which allows "stacking", btw) maintain that lookup?

**Comment by mvaldes14** [comment]
+1 on having a blocked version, as someone that makes small plugins and utilities that rely on selection and doing something else with the nested callback is problematic so if we could pause the execution until something is picked by a user it would facilitate the flow of information.

This is specifically useful when a selection spawns another ui element like a window or input. By default everything launches and becomes nil since it doesn't wait.

**Comment by justinmk** [comment]
~~A caller to `vim.ui.select()` can just wrap it with `vim.wait()`~~. Edit: no, that is too limited (can't run terminal, etc).

## Proposal

see comments starting at https://github.com/neovim/neovim/issues/24632#issuecomment-1674408050

**Comment by ymic9963** [comment]
I'd like to work on this since I need it for my plugin, but I'm trying to understand the issue. The default behaviour is already blocking, so what's the issue exactly? Is it because the callback gets treated asynchronously in some cases?

**Comment by justinmk** [comment]
`vim.ui.select()` has a non-blocking contract. Even if the default happens to work, that doesn't help when users override it with e.g. fzf-lua.

> I need it for my plugin, ... what's the issue exactly?

I'm confused 😅

One way to solve this is https://github.com/neovim/neovim/pull/40463 , which works by spawning a child `nvim` so that the child can be fully asynchronous and the parent can block while the child does whatever it wants. The tradeoff is that the child doesn't have all the global state of the parent...

**Comment by ymic9963** [comment]
> I'm confused 😅

I've been trying bugfix for a while so I'm pretty fried...please ignore any nonsense I type. 

Regarding the issue, forgive my lack of knowledge but doesn't it make sense for a selection to always have it be blocking/synchronous? Then you could have it be async with a `coroutine`. Am I talking nonsense again?

Another thing that's also confusing me is that this code with default behaviour works, but overriding it (in the case of fzf-lua) doesn't work?
```lua
    local index = nil
    vim.ui.select({"test1", "test"}, {
        prompt = "Select:",
    }, function (_, idx)
        index = idx
    end)
    vim.print(index)
```
Here `index` will always print `nil` when overriding, however it will print the selection index with the default behaviour. Why?

**Comment by justinmk** [comment]
was discussed above and in linked issues. does not address this issue.
